scipy/signal/ltisys.py

```diff
@@ -658,58 +660,101 @@ def lsim(system, U, T, X0=None, interp=1):
         sys = system
     else:
         sys = lti(*system)
-    U = atleast_1d(U)
     T = atleast_1d(T)
-    if len(U.shape) == 1:
-        U = U.reshape((U.shape[0], 1))
-    sU = U.shape
     if len(T.shape) != 1:
         raise ValueError("T must be a rank-1 array.")
-    if sU[0] != len(T):
-        raise ValueError("U must have the same number of rows "
-                         "as elements in T.")
-    if sU[1] != sys.inputs:
-        raise ValueError("System does not define that many inputs.")
 
+    A, B, C, D = (np.asarray(M) for M in (sys.A, sys.B, sys.C, sys.D))
+    n_states = A.shape[0]
+    n_inputs = B.shape[1]
+
+    n_steps = len(T)
     if X0 is None:
-        X0 = zeros(sys.B.shape[0], sys.A.dtype)
+        X0 = zeros(n_states, sys.A.dtype)
+    xout = zeros((n_steps, n_states), sys.A.dtype)
+
+    if T[0] == 0:
+        xout[0] = X0
+    elif T[0] > 0:
+        # step forward to initial time, with zero input
+        xout[0] = dot(X0, linalg.expm(transpose(A) * T[0]))
+    else:
+        raise ValueError("Initial time must be nonnegative")
+
+    no_input = (U is None
+                or (isinstance(U, (int, float)) and U == 0.)
+                or np.allclose(U, 0))
```

Comment:
This ``allclose`` usage looks wrong:

    In [1]: np.allclose([0, 1e-9, 2e-9], 0)
    Out[1]: True